src/main/java/org/opensearch/knn/index/KNNIndexShard.java

```diff
@@ -86,42 +81,183 @@ public String getIndexName() {
+         */
+        boolean warmUp(FieldInfo field) throws IOException;
+
+        default boolean checkedWarmUp(FieldInfo field) {
+            try {
+                return warmUp(field);
+            } catch (IOException e) {
+                log.error("Failed to warm up field: {}: {}", field.getName(), e.toString());
                 return false;
-            })
-            .collect(Collectors.toSet());
+            }
+        }
+    }
 
-        final SegmentReader segmentReader = Lucene.segmentReader(leafReader);
-        for (final FieldInfo field : fieldsForMemoryOptimizedSearch) {
+    /**
+     * Partially warm up the index by doing a search with an empty query vector
+     */
+    private class PartialFieldWarmUpStrategy implements FieldWarmUpStrategy {
+        SegmentReader segmentReader;
+        LeafReader leafReader;
+
+        PartialFieldWarmUpStrategy(LeafReader leafReader) {
+            this.leafReader = leafReader;
+            this.segmentReader = Lucene.segmentReader(leafReader);
+        }
+
+        @Override
+        public boolean warmUp(FieldInfo field) throws IOException {
             final String dataTypeStr = field.getAttribute(VECTOR_DATA_TYPE_FIELD);
+
             if (dataTypeStr == null) {
-                continue;
+                return false;
             }
-            try {
-                // Partial load Faiss index by triggering search.
-                final VectorDataType vectorDataType = VectorDataType.get(dataTypeStr);
-                if (vectorDataType == VectorDataType.FLOAT) {
-                    segmentReader.getVectorReader().search(field.getName(), (float[]) null, null, null);
-                } else {
-                    segmentReader.getVectorReader().search(field.getName(), (byte[]) null, null, null);
-                }
-            } catch (Exception e) {
-                // Ignore
+
+            final VectorDataType vectorDataType = VectorDataType.get(dataTypeStr);
+            if (vectorDataType == VectorDataType.FLOAT) {
+                segmentReader.getVectorReader().search(field.getName(), (float[]) null, null, null);
+            } else {
+                segmentReader.getVectorReader().search(field.getName(), (byte[]) null, null, null);
             }
+            return true;
         }
+    }
+
+    /**
+     * Fully warm up the index by loading every byte from disk, causing page faults
+     */
+    private class FullFieldWarmUpStrategy implements FieldWarmUpStrategy {
+        SegmentReader segmentReader;
+        Directory directory;
+        LeafReader leafReader;
+
+        FullFieldWarmUpStrategy(LeafReader leafReader, Directory directory) {
+            this.leafReader = leafReader;
+            this.segmentReader = Lucene.segmentReader(leafReader);
+            this.directory = directory;
+        }
+
+        private void pageFaultFile(String file) throws IOException {
+            IndexInput input = directory.openInput(file, IOContext.DEFAULT);
```

Comment:
- Lets use `IOContext.READONCE` to avoid any sideaffects
- Also can we wrap this up in try with resources to make sure file is unmapped?

Reply:
```suggestion
            try (IndexInput input = directory.openInput(file, IOContext.READONCE)) {
                      // logic here
            }
```
This takes of resource closure in case of error cases. It makes sure file is unmapped in case an error is thrown